internal/clierror/error.go

```diff
@@ -1,22 +1,152 @@
 package clierror
 
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/maruel/panicparse/v2/stack"
+	log "github.com/sirupsen/logrus"
+)
+
 // SanitizedError allows errors to be wrapped with a sanitized message for sending upstream
-type SanitizedError struct {
+type SanitizedError interface {
+	SanitizedError() string
+	SanitizedStack() string
+}
+
+type CLIError struct {
 	sanitizedMsg string
 	err          error
 }
 
-func (e *SanitizedError) Error() string {
+func NewCLIError(err error, sanitizedMsg string) *CLIError {
+	return &CLIError{
+		sanitizedMsg: sanitizedMsg,
+		err:          err,
+	}
+}
+
+func (e *CLIError) Error() string {
 	return e.err.Error()
 }
 
-func (e *SanitizedError) SanitizedError() string {
+func (e *CLIError) SanitizedError() string {
 	return e.sanitizedMsg
 }
 
-func NewSanitizedError(err error, sanitizedMsg string) *SanitizedError {
-	return &SanitizedError{
-		sanitizedMsg: sanitizedMsg,
-		err:          err,
+func (e *CLIError) SanitizedStack() string {
+	return ""
+}
+
+// PanicError is used to collect goroutine panics into an error interface so
+// that we can do type assertion on err checking.
+type PanicError struct {
+	err   error
+	stack []byte
+}
+
+func NewPanicError(err error, stack []byte) *PanicError {
+	return &PanicError{
+		err:   err,
+		stack: stack,
 	}
 }
+
+func (p *PanicError) Error() string {
+	return fmt.Sprintf("%s\n%s", p.err.Error(), p.stack)
+}
+
+func (p *PanicError) SanitizedError() string {
+	return p.err.Error()
+}
+
+func (p *PanicError) SanitizedStack() string {
+	sanitizedStack := p.stack
+	sanitizedStack, err := processStack(sanitizedStack)
+	if err != nil {
+		log.Debugf("Could not sanitize stack: %s", err)
+	}
+
+	return string(sanitizedStack)
+}
+
+// processStack processes the raw stack trace into a format that can be reported upstream.
+// This strips out the absolute path information of the local machine and removes any function args
+// so that the reported value always equals the same string value as an equivalent stack trace.
+// Adapted from: https://pkg.go.dev/github.com/maruel/panicparse/v2/stack#example-package-Text
+func processStack(rawStack []byte) ([]byte, error) {
```

Comment:
can we add a defer panic catch in this function since this does a lot and it's designed to handle panics. We don't want to do all this work and then just not catch a panic at this level!

Reply:
Discussed in person. We could change this so it just sends the raw stack trace if we catch a panic here, but decided it wasn't worth doing just now.